# Ebay Kleinanzeigen Dataset for Data cleaning Exercise

The dataset was scraped and uploaded to Kaggle. In this exercise, the dataset had been modified as follows:
1. Only 50,000 data points were randomly smapled for the speed.
2. A dirtied version of the dataset to reflect the original status of a scraped dataset.


In [6]:
!pip install numpy
!pip install pandas

In [109]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [110]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Observation 

After importing to a dataframe: autos, we noticed a few things about this dataset. 
1. There are 20 columns in total, including 5 columns in numeric format and others a mixture of dates, strings and a combination of integers and strings.
2. 15 columns have 50,000 entries while 5 columns have missing entries, null values ranging from (0, 10,000)
3. This dataset takes more than 7.6MB of the memory
4. Some column names contain camelcase instead of using Python's snakecase convention.

Therefore, let's change the column names to better understanding snakecases.

In [111]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [112]:
for c in autos.columns:
    if c == "yearOfRegistration":
        autos.rename({"yearOfRegistration":"registration_year"},inplace=True, axis=1)
    elif c == "monthOfRegistration":
        autos.rename({"monthOfRegistration":"registration_month"},inplace=True, axis=1)
    elif c == "notRepairedDamage":
        autos.rename({"notRepairedDamage":"unrepaired_damage"},inplace=True, axis=1)
    elif c == "dateCreated":
        autos.rename({"dateCreated":"ad_created"},inplace=True, axis=1)
    else:     
        autos.rename({c:c.lower()},inplace=True, axis=1)   

### Clean column names

In the above code block, we have changed the camelcased column names to snakecased while modified four of them to shorter and simpler names.

In [113]:
autos.describe(include="all")

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [114]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [115]:
autos["offertype"].value_counts()

Angebot    49999
Gesuch         1
Name: offertype, dtype: int64

In [116]:
autos["nrofpictures"].value_counts()

0    50000
Name: nrofpictures, dtype: int64

In [117]:
autos["unrepaired_damage"].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

In [118]:
autos["gearbox"].value_counts()

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

Change the German words in column "gearbox", "unrepaired_damage" to English

In [119]:
mapping_directory = {"manuell": "manual", "automatik": "automatic"}
autos["gearbox"] = autos['gearbox'].map(mapping_directory)

mapping_directory = {"nein": "no", "ja": "yes"}
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(mapping_directory)

In [120]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0     manual      158  andere  150,000km                   3      lpg   
1  automatic      286     7er  150,000km                   6   benzin   
2     manual      102    golf   70,000km                   7   benzin   
3  automatic       71  fortwo   70,000km                   6   benzin   
4     manual        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot                no  2016-03-26 00:00:00             0   
1         bmw                no  2016-04-04 00:00:00             0   
2  volkswagen                no  2016-03-26 00:00:00             0   
3       smart                no  2016-03-12 00:00:00             0   
4        ford                no  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Potential Cleaning

After investigating the descriptive information of the dataset, we have detected a few potential cleaning tasks to be completed.
1. column "seller" and "offertype" both have almost one value. Probably not very meaningful for our analysis. These two columns could be dropped. "nrofpictures" has a mono value of 50000 is not very useful and to be dropped.
2. both "price" and "odometer" columns are essentially numbers but disguised in strings. We should turn them into floats for further analysis.
3. "datecrawled" is by seconds which probably are too detailed to be used for our purpose. Potentially, we could turn this into just year-month or year-month-day or just consider them the same and drop this column
4. Obviously there are some mistakes in the column "registration_year". Most likely any number beyond (1900, 2016) is an error and should be marked as "unknown" or "nan". Meanwhile, column "registration_month" has 0s which also requires attention. 
5. Similarly, column "powerps" being 0 is probably an error and need to be adjusted. 

Let's first transfer the strings to integers for "price" and "odometer".

In [121]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)

In [122]:
autos["odometer"]=autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [123]:
autos.rename({"odometer":"odometer_km"},inplace=True, axis=1)

In [124]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  5000.0  control         bus               2004   
1  privat   Angebot  8500.0  control   limousine               1997   
2  privat   Angebot  8990.0     test   limousine               2009   
3  privat   Angebot  4350.0  control  kleinwagen               2007   
4  privat   Angebot  1350.0     test       kombi               2003   

     gearbox  powerps   model  odometer_km  registration_month fueltype  \
0     manual      158  andere       150000                   3      lpg   
1  automatic      286     7er       150000                   6   benzin   
2     manual      102    golf        70000                   7   benzin   
3  automatic       71  fortwo        70000                   6   benzin   
4     manual        0   focus       150000                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot                no  2016-03-26 00:00:00             0   
1         bmw                no  2016-04-04 00:00:00             0   
2  volkswagen                no  2016-03-26 00:00:00             0   
3       smart                no  2016-03-12 00:00:00             0   
4        ford                no  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [125]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [126]:
autos["price"].value_counts().sort_index().head(50)


0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
8.0         1
9.0         1
10.0        7
11.0        2
12.0        3
13.0        2
14.0        1
15.0        2
17.0        3
18.0        1
20.0        4
25.0        5
29.0        1
30.0        7
35.0        1
40.0        6
45.0        4
47.0        1
49.0        4
50.0       49
55.0        2
59.0        1
60.0        9
65.0        5
66.0        1
70.0       10
75.0        5
79.0        1
80.0       15
89.0        1
90.0        5
99.0       19
100.0     134
110.0       3
111.0       2
115.0       2
117.0       1
120.0      39
122.0       1
125.0       8
129.0       1
130.0      15
135.0       1
139.0       1
140.0       9
Name: price, dtype: int64

In [127]:
autos.loc[autos["price"]>10000000, :]

datecrawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   

       seller offertype       price   abtest vehicletype  registration_year  \
2897   privat   Angebot  11111111.0     test   limousine               1973   
24384  privat   Angebot  11111111.0     test         NaN               1995   
27371  privat   Angebot  12345678.0  control         NaN               2017   
39377  privat   Angebot  12345678.0  control         NaN               2018   
39705  privat   Angebot  99999999.0  control   limousine               1999   
42221  privat   Angebot  27322222.0  control   limousine               2014   
47598  privat   Angebot  12345678.0  control   limousine               2001   

         gearbox  powerps     model  odometer_km  registration_month fueltype  \
2897      manual       48    escort        50000                   3   benzin   
24384        NaN        0       NaN       150000                   0      NaN   
27371        NaN       95     punto       150000                   0      NaN   
39377     manual       95       v40       150000                   6      NaN   
39705  automatic      224  s_klasse       150000                   9   benzin   
42221     manual      163        c4        40000                   2   diesel   
47598     manual      101    vectra       150000                   3   benzin   

               brand unrepaired_damage           ad_created  nrofpictures  \
2897            ford                no  2016-03-12 00:00:00             0   
24384     volkswagen               NaN  2016-03-21 00:00:00             0   
27371           fiat               NaN  2016-03-09 00:00:00             0   
39377          volvo                no  2016-03-08 00:00:00             0   
39705  mercedes_benz               NaN  2016-03-22 00:00:00             0   
42221        citroen               NaN  2016-03-08 00:00:00             0   
47598           opel                no  2016-03-31 00:00:00             0   

       postalcode             lastseen  
2897        94469  2016-03-12 22:45:27  
24384       18519  2016-03-21 14:40:18  
27371       96110  2016-03-09 15:45:47  
39377       14542  2016-04-06 23:17:31  
39705       73525  2016-04-06 05:15:30  
42221       76532  2016-03-08 20:39:05  
47598        4356  2016-03-31 18:56:54

In [128]:
autos.drop(autos[autos["price"] > 10000000].index, inplace=True)

In [129]:
autos.loc[autos["price"]>10000000, :]

Empty DataFrame
Columns: [datecrawled, name, seller, offertype, price, abtest, vehicletype, registration_year, gearbox, powerps, model, odometer_km, registration_month, fueltype, brand, unrepaired_damage, ad_created, nrofpictures, postalcode, lastseen]
Index: []

Drop the two columns "seller" and "offertype"

In [130]:
autos.drop(autos.loc[:,["seller", "offertype"]], inplace = True, axis = 1)

In [131]:
autos.drop(autos.loc[:,["nrofpictures"]], inplace = True, axis = 1)

In [132]:
autos.head(40)

datecrawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20  2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21  2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22  2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23  2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24  2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25  2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26  2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27  2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28  2016-03-19 21:56:19                                      MINI_Cooper_D   
29  2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30  2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
31  2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
32  2016-03-20 05:03:03                               Corsa_mit_TÜV_5.2016   
33  2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34  2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35  2016-03-29 21:52:56                                        Ford_Mondeo   
36  2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37  2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38  2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39  2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   

      price   abtest vehicletype  registration_year    gearbox  powerps  \
0    5000.0  control         bus               2004     manual      158   
1    8500.0  control   limousine               1997  automatic      286   
2    8990.0     test   limousine               2009     manual      102   
3    4350.0  control  kleinwagen               2007  automatic       71   
4    1350.0     test       kombi               2003     manual        0   
5    7900.0     test         bus               2006  automatic      150   
6     300.0     test   limousine               1995     manual       90   
7    1990.0  control   limousine               1998     manual       90   
8     250.0     test         NaN               2000     manual        0   
9     590.0  control         bus               1997     manual 

### Dropping rows based on "Price"

Some vehicles have price tags more than 10million and the numbers look artibary, e.g. 12345678. It seems like those are dummy inputs in the system. Therefore, we have decided to delete those 5 records.

Next we turn to the date columns


In [133]:
autos['datecrawled'] = autos['datecrawled'].str[:10]

In [134]:
autos['ad_created'] = autos['ad_created'].str[:10]

In [135]:
autos['lastseen'] = autos['lastseen'].str[:10]




Using the string method above, we have successfully taking the year-month-day for the three date columns for easy analysis.

In [136]:
autos['lastseen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005361
2016-03-08    0.007581
2016-03-09    0.009841
2016-03-10    0.010762
2016-03-11    0.012522
2016-03-12    0.023803
2016-03-13    0.008981
2016-03-14    0.012802
2016-03-15    0.015882
2016-03-16    0.016442
2016-03-17    0.027924
2016-03-18    0.007421
2016-03-19    0.015742
2016-03-20    0.020703
2016-03-21    0.020723
2016-03-22    0.021583
2016-03-23    0.018583
2016-03-24    0.019563
2016-03-25    0.019203
2016-03-26    0.016962
2016-03-27    0.016022
2016-03-28    0.020863
2016-03-29    0.022343
2016-03-30    0.024843
2016-03-31    0.023823
2016-04-01    0.023103
2016-04-02    0.024903
2016-04-03    0.025364
2016-04-04    0.024623
2016-04-05    0.124297
2016-04-06    0.220991
2016-04-07    0.130938
Name: lastseen, dtype: float64

"lastseen" column has been created by the crawler in a month time, ranges from 5, March, 2016 to 7, April, 2016.

In [137]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033225
2016-03-10    0.031864
2016-03-11    0.032785
2016-03-12    0.036605
2016-03-13    0.016922
2016-03-14    0.035225
2016-03-15    0.033745
2016-03-16    0.030004
2016-03-17    0.031204
2016-03-18    0.013722
2016-03-19    0.033845
2016-03-20    0.037865
2016-03-21 

Most of the ads were created in the period of crawling.

In [138]:
autos['datecrawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025384
2016-03-06    0.013942
2016-03-07    0.035965
2016-03-08    0.033265
2016-03-09    0.033205
2016-03-10    0.032124
2016-03-11    0.032485
2016-03-12    0.036765
2016-03-13    0.015562
2016-03-14    0.036625
2016-03-15    0.033985
2016-03-16    0.029504
2016-03-17    0.031524
2016-03-18    0.013062
2016-03-19    0.034905
2016-03-20    0.037825
2016-03-21    0.037505
2016-03-22    0.032925
2016-03-23    0.032385
2016-03-24    0.029104
2016-03-25    0.031744
2016-03-26    0.032485
2016-03-27    0.031044
2016-03-28    0.034845
2016-03-29    0.034185
2016-03-30    0.033625
2016-03-31    0.031904
2016-04-01    0.033805
2016-04-02    0.035405
2016-04-03    0.038685
2016-04-04    0.036525
2016-04-05    0.013102
2016-04-06    0.003180
2016-04-07    0.001420
Name: datecrawled, dtype: float64

The crawling happened during that same month and rather evenly distributed in that period.

In [139]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

1000    0.000020
1001    0.000020
1111    0.000020
1500    0.000020
1800    0.000040
1910    0.000180
1927    0.000020
1929    0.000020
1931    0.000020
1934    0.000040
1937    0.000080
1938    0.000020
1939    0.000020
1941    0.000040
1943    0.000020
1948    0.000020
1950    0.000060
1951    0.000040
1952    0.000020
1953    0.000020
1954    0.000040
1955    0.000040
1956    0.000100
1957    0.000040
1958    0.000080
1959    0.000140
1960    0.000680
1961    0.000120
1962    0.000080
1963    0.000180
          ...   
2001    0.054048
2002    0.050667
2003    0.054548
2004    0.054748
2005    0.060308
2006    0.054168
2007    0.046086
2008    0.044626
2009    0.041966
2010    0.031944
2011    0.032685
2012    0.026464
2013    0.016122
2014    0.013302
2015    0.007981
2016    0.026324
2017    0.029044
2018    0.009821
2019    0.000060
2800    0.000020
4100    0.000020
4500    0.000020
4800    0.000020
5000    0.000080
5911    0.000020
6200    0.000020
8888    0.000020
9000    0.0000

Clearly that some of the years in the list are not valid. Since the dataset was generated in 2016, any number beyong 2016 should be removed. More the car was invented in 1885. Therefore, we can savely remove all records before 1900.

In [140]:
autos.drop(autos[(autos["registration_year"]>2016) | (autos["registration_year"]<1900)].index, inplace=True)

In [141]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000708
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003769
1990    0.008225
1991    0.007413
1992    0.008142
1993    0.009266
1994    0.013743
1995    0.027320
1996    0.030069
1997    0.042230
1998    0.051080
1999    0.062449
2000    0.069842
2001    0.056265
2002    0.052746
2003    0.056785
2004    0.056994
2005    0.062782
2006    0.056390
2007    0.047977
2008    0.046457
2009    0.043687
2010    0.033255
2011    0.034025
2012    0.027549
2013    0.016784
2014    0.0138

Most of the car in the dataset was registered after the 90s.

In [142]:
autos["brand"].value_counts(normalize=True, dropna=False)

volkswagen        0.212128
bmw               0.110031
opel              0.108157
mercedes_benz     0.095350
audi              0.086396
ford              0.069779
renault           0.047352
peugeot           0.029528
fiat              0.025863
seat              0.018179
skoda             0.016034
mazda             0.015139
nissan            0.015097
smart             0.013910
citroen           0.013910
toyota            0.012473
sonstige_autos    0.010953
hyundai           0.009849
volvo             0.009246
mini              0.008642
mitsubishi        0.008142
honda             0.007850
kia               0.007101
alfa_romeo        0.006622
porsche           0.006101
suzuki            0.005914
chevrolet         0.005706
chrysler          0.003665
daihatsu          0.002561
dacia             0.002561
jeep              0.002249
subaru            0.002186
land_rover        0.002041
saab              0.001603
jaguar            0.001583
trabant           0.001562
daewoo            0.001499
r

In [143]:
autos["brand"].value_counts(normalize=True, dropna=False)

volkswagen        0.212128
bmw               0.110031
opel              0.108157
mercedes_benz     0.095350
audi              0.086396
ford              0.069779
renault           0.047352
peugeot           0.029528
fiat              0.025863
seat              0.018179
skoda             0.016034
mazda             0.015139
nissan            0.015097
smart             0.013910
citroen           0.013910
toyota            0.012473
sonstige_autos    0.010953
hyundai           0.009849
volvo             0.009246
mini              0.008642
mitsubishi        0.008142
honda             0.007850
kia               0.007101
alfa_romeo        0.006622
porsche           0.006101
suzuki            0.005914
chevrolet         0.005706
chrysler          0.003665
daihatsu          0.002561
dacia             0.002561
jeep              0.002249
subaru            0.002186
land_rover        0.002041
saab              0.001603
jaguar            0.001583
trabant           0.001562
daewoo            0.001499
r

The top 3 brands cover more than 40% of the total cars. Therefore, we will aggregate based on the top three brands.

In [144]:
brand_price = {}
brand_list = ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'subaru',
       'land_rover', 'saab', 'jaguar', 'trabant', 'daewoo', 'rover', 'lancia',
       'lada']

for index, row in autos.iterrows():
    b = row["brand"]
    if b in brand_list: 
        if b in brand_price:
            brand_price[b] += row["price"]
        else:
            brand_price[b] = row["price"]


In [145]:
brand_count = {}

for index, row in autos.iterrows():
    b = row["brand"]
    if b in brand_list: 
        if b in brand_count:
            brand_count[b] += 1
        else:
            brand_count[b] = 1
    else:
        if "others" in brand_count:
            brand_count["others"] += 1
        else:
            brand_count["others"] = 1

In [146]:
brand_avg_price = {}
for brand, price in brand_price.items():
    for k, v in brand_count.items():
        if brand == k:
            brand_avg_price[brand]=price/v
brand_sort = []
for brand, avg_price in brand_avg_price.items():
    brand_sort.append([avg_price, brand])
brand_sort.sort(reverse=True)

for item in brand_sort:
    print(f'Average price of {item[1].upper()}: {item[0]:.2f} US$')


Average price of PORSCHE: 44553.47 US$
Average price of SONSTIGE_AUTOS: 39621.78 US$
Average price of LAND_ROVER: 19108.09 US$
Average price of JEEP: 11434.75 US$
Average price of JAGUAR: 11176.20 US$
Average price of MINI: 10460.01 US$
Average price of AUDI: 9093.65 US$
Average price of MERCEDES_BENZ: 8485.24 US$
Average price of BMW: 8334.65 US$
Average price of CHEVROLET: 6488.98 US$
Average price of SKODA: 6334.92 US$
Average price of DACIA: 5915.53 US$
Average price of KIA: 5789.35 US$
Average price of VOLKSWAGEN: 5426.38 US$
Average price of HYUNDAI: 5308.54 US$
Average price of TOYOTA: 5115.33 US$
Average price of VOLVO: 4757.11 US$
Average price of NISSAN: 4664.89 US$
Average price of SEAT: 4296.49 US$
Average price of MAZDA: 4010.77 US$
Average price of SUZUKI: 3995.76 US$
Average price of HONDA: 3988.00 US$
Average price of ALFA_ROMEO: 3984.86 US$
Average price of FORD: 3949.42 US$
Average price of SUBARU: 3765.04 US$
Average price of CITROEN: 3699.94 US$
Average price of SMA

Based on the aggregation of the average prices on brands, AUDI has the highest mean price, followed by another two German brands: BENZ and BMW. The top three selling brands are a mix of mid-range VOLKSWAGAN, lower-tier OPEL and the upper class BMW. 

Next, we calculate the average miles per brand to study if there is any correlation between the price and the mileage.

In [147]:
brand_mile = {}

for index, row in autos.iterrows():
    b = row["brand"]
    if b in brand_list: 
        if b in brand_mile:
            brand_mile[b] += row["odometer_km"]
        else:
            brand_mile[b] = row["odometer_km"]
print(brand_mile)

{'peugeot': 180280000, 'bmw': 699785000, 'volkswagen': 1311355000, 'smart': 66530000, 'ford': 415755000, 'chrysler': 23440000, 'seat': 106125000, 'renault': 291490000, 'mercedes_benz': 599190000, 'audi': 536415000, 'sonstige_autos': 45900000, 'opel': 671185000, 'mazda': 90690000, 'porsche': 28555000, 'mini': 36770000, 'toyota': 69310000, 'dacia': 10365000, 'nissan': 85965000, 'jeep': 13775000, 'saab': 11120000, 'volvo': 61430000, 'mitsubishi': 49630000, 'jaguar': 9190000, 'fiat': 144760000, 'skoda': 85435000, 'subaru': 13110000, 'kia': 38340000, 'citroen': 79880000, 'chevrolet': 27195000, 'hyundai': 50380000, 'honda': 46315000, 'daewoo': 8815000, 'suzuki': 30970000, 'trabant': 4475000, 'land_rover': 11565000, 'alfa_romeo': 41785000, 'lada': 2480000, 'rover': 8815000, 'daihatsu': 14180000, 'lancia': 6345000}


In [148]:
brand_avg_mile = {}
for brand, mile in brand_mile.items():
    for k, v in brand_count.items():
        if brand == k:
            brand_avg_mile[brand]=mile/v
brand_sort = []
for brand, avg_mile in brand_avg_mile.items():
    brand_sort.append([avg_mile, brand])
brand_sort.sort(reverse=True)

for item in brand_sort:
    print(f'Average mileage of {item[1].upper()}: {item[0]:.2f} KM')

Average mileage of SAAB: 144415.58 KM
Average mileage of VOLVO: 138355.86 KM
Average mileage of ROVER: 135615.38 KM
Average mileage of CHRYSLER: 133181.82 KM
Average mileage of BMW: 132434.71 KM
Average mileage of ALFA_ROMEO: 131399.37 KM
Average mileage of MERCEDES_BENZ: 130856.08 KM
Average mileage of AUDI: 129287.78 KM
Average mileage of OPEL: 129223.14 KM
Average mileage of VOLKSWAGEN: 128728.28 KM
Average mileage of RENAULT: 128183.82 KM
Average mileage of JEEP: 127546.30 KM
Average mileage of PEUGEOT: 127136.81 KM
Average mileage of MITSUBISHI: 126930.95 KM
Average mileage of SUBARU: 124857.14 KM
Average mileage of MAZDA: 124745.53 KM
Average mileage of FORD: 124068.93 KM
Average mileage of HONDA: 122851.46 KM
Average mileage of DAEWOO: 122430.56 KM
Average mileage of LANCIA: 122019.23 KM
Average mileage of SEAT: 121563.57 KM
Average mileage of JAGUAR: 120921.05 KM
Average mileage of CITROEN: 119580.84 KM
Average mileage of NISSAN: 118572.41 KM
Average mileage of LAND_ROVER: 1180

In [149]:
# convert the dictionaries of the mean_price and mean_mileage to df
bmp_series = pd.Series(brand_avg_price)
bmm_series = pd.Series(brand_avg_mile)
freq_series = pd.Series(autos["brand"].value_counts(normalize=True, dropna=False))

In [150]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df["mean_mileage"] = bmm_series
df["count_perc"] = freq_series

In [151]:
print(df.sort_index())

                  mean_price   mean_mileage  count_perc
alfa_romeo       3984.855346  131399.371069    0.006622
audi             9093.650036  129287.780188    0.086396
bmw              8334.645155  132434.708554    0.110031
chevrolet        6488.981752   99251.824818    0.005706
chrysler         3229.443182  133181.818182    0.003665
citroen          3699.935629  119580.838323    0.013910
dacia            5915.528455   84268.292683    0.002561
daewoo           1019.861111  122430.555556    0.001499
daihatsu         1556.382114  115284.552846    0.002561
fiat             2711.801127  116553.945250    0.025863
ford             3949.423456  124068.934646    0.069779
honda            3988.000000  122851.458886    0.007850
hyundai          5308.539112  106511.627907    0.009849
jaguar          11176.197368  120921.052632    0.001583
jeep            11434.750000  127546.296296    0.002249
kia              5789.351906  112434.017595    0.007101
lada             2502.896552   85517.241379    0

In [152]:
autos.head()

datecrawled                                               name   price  \
0  2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  5000.0   
1  2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  8500.0   
2  2016-03-26                         Volkswagen_Golf_1.6_United  8990.0   
3  2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  4350.0   
4  2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  1350.0   

    abtest vehicletype  registration_year    gearbox  powerps   model  \
0  control         bus               2004     manual      158  andere   
1  control   limousine               1997  automatic      286     7er   
2     test   limousine               2009     manual      102    golf   
3  control  kleinwagen               2007  automatic       71  fortwo   
4     test       kombi               2003     manual        0   focus   

   odometer_km  registration_month fueltype       brand unrepaired_damage  \
0       150000                   3      lpg     peugeot                no   
1       150000                   6   benzin         bmw                no   
2        70000                   7   benzin  volkswagen                no   
3        70000                   6   benzin       smart                no   
4       150000                   7   benzin        ford                no   

   ad_created  postalcode    lastseen  
0  2016-03-26       79588  2016-04-06  
1  2016-04-04       71034  2016-04-06  
2  2016-03-26       35394  2016-04-06  
3  2016-03-12       33729  2016-03-15  
4  2016-04-01       39218  2016-04-01

In [153]:
autos["num_years"]=2016-autos["registration_year"]

In [154]:
df = pd.DataFrame(autos.loc[:,["name","num_years","odometer_km"]])

In [155]:
df.sort_values(by=['odometer_km'])

name  num_years  \
6543   2×_Seat_toledo_1_9tdi_mit_klima_fuer_1700euro_...         11   
43481                              Holzkoepfe/Arschmaden         16   
43492         Opel_Astra_G_mit_Klima_elektrische_fenster         16   
3868                                   Skoda_Felicia_1_6         16   
36905  Verkaufe_den_guterhaltenen_Opel_signum__opc_3_...         11   
26309                      BMW_320_I_W3_Sportausstattung         18   
36892                          Audi_A3_1.2_TFSI_Ambition          0   
20180  Renault_Clio_Grandtour_Energy_TCe_120_EDC_Limited          0   
43514            Verkaufe_Golf_2_cl_in_org_gruen_TÜV_NEU         26   
26298                                 Opel_Adam_1.4_Slam          1   
26291     Skoda_Citigo_MPI_Automatik_Ambition___5_tuerig          2   
29320                   Audi_a4___2._5_l_gruene__plakete         12   
3928   Oldtimer_GAZ_M_21_Wolga_viele_Ersatzteile_Youn...         56   
16255  ALFA_ROMEO_156_Twin_Spark_2.0_16v_mit_tuev_11/...         17   
9951                           Der_Sommer_kommt_Open_Air         15   
47493                               Mercedes_C_Clase_220         14   
3949                                   Volkswagen_Kaefer         41   
26354                       Motorsport_Corsa_B_ca.120_Ps         23   
43449              An_Bastler_Schrauber_oder_Export.....         18   
26468                                        Kaufe_Autos         16   
20060                                        Opel_vectra         12   
32981  Volkswagen_T6_California_Ocean_DSG/4Motion__SO...          0   
1011                                    Ford_top_Zustand         17   
39425  Mercedes_Benz_C_200_Avantgarde//Exclusive_NEU_...          0   
20083                                Fiat_punto_teile!!!         11   
33014                               Verkaufe_Bastlerauto         16   
36982                                Lpg_auto_und_benzin         16   
43440  Audi_A6_3.0_V6_TDI_Avant_Quattro_Navi_Leder__V...         11   
33008                                       BMW_limosine         16   
3796                                        Volvo_Amazon         50   
...                                                  ...        ...   
20174          Smart_forfour__1.1_passion_8_fach_bereift          9   
20166  Mitsubishi_Space_Star_1.6__Klima_Navi___TOP_TÜ...         12   
20175                      Mercedes_250_TD__W_124__Kombi         24   
20177             E_klasse_taxi_sauber_motor_getriebe_1a          9   
20179                    Mercedes_Benz_C_220_T_CDI_Sport         17   
20182                                   BMW_525d_Touring         13   
20183                                   Chrysler_Voyager         20   
20186  BMW_530d_touring_M_Paket_/_Xenon_/_Leder_/_Nav...         14   
20188          VW_Golf_GTI_1.6_4_gang_1979_seltene_Farbe         37   
20176                                        BMW_X3_3.0d         12   
20165                                      Audi_80_2.0_E         23   
20164                                Mercedes_Benz_190_E         26   
20163  VW_PASSAT_KOMBI_AUTOMATIK_KLIMAAUTOMATIK_ALUFE...         19   
20139                                     Opel_Combo_1.7         14   
20140                              Roomster_1_9_TDI_2006         10   
20141           2006_FORD_FOCUS_KOMBI_1_6_TDCI_AUTOMATIK         10   
20143                                Ssangyong_Musso_E23         19   
20144     Mercedes_Benz_C_220_CDI_Sportcoupe_aus_1._Hand         15   
20145                                         Ford_Focus         16   
20147   Golf_4_1.9_Tdi_Pacific_Limousine_gruene_Plakette         13   
20148      Opel_GTC_Sport_1.8_OPC_Recaro_Sitzheizung_Alu          7   
20150                                   BMW_523i_touring         16   
20152                                  Volkswagen_Beetle         17   
20153                     Honda_Civic_Coupe_1.7_LS_Klima         15   
20156                                Mercedes_Benz_A_160         14   
20159